### PR TITLE
feat(CODEOWNERS): Improve pr files owner detection

### PR DESCRIPTION
**Pull request by lluiscab**
## Changes

This PR improves how the CODEOWNERS file is parsed & matched against the modified files in the PR.

## Context

The current implementation of CODEOWNERS leaves a lot to be desired, my coworkers and I have found that the current implementation does not match what's seen in VSCode using the [vscode-codeowners](https://marketplace.visualstudio.com/items?itemName=jasonnutter.vscode-codeowners) extension and that when a merge requests modifies multiple files in a monorepo (for example, multiple package.json files with different CODEOWNERS assigned), they get assigned the wrong reviewer as the current implementation does not take `priority` into account

I've changed the implementation in such a way that it now assigns a `score` to each matched user for each modified file and then sorts the users based on said score. This ensures that wildcard matches are less prioritized than exact matches

### Example 1

Given the following CODEOWNERS file
```
* @john
packages/a/ @maria
```

and the following changes files:
 - package.json
 - packages/a/package.json

The result will be
 - @maria (from packages/a)
 - @john (from the wildcard)

### Example 2

This changes also fix a issue found when a wildcard is present but some files have been declared as orphans.

Given the following CODEOWNERS file
```
* @john
yarn.lock
```

and the following changes files:
 - yarn.lock
 
This will now result in no users being identified as owners of any changes.

More examples can be seen by reviewing the newly added tests

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
I think we need to check if there is "one true way" to do codeowners logic, and if this is it. If the way to do it is opinionated - and the vscode extension is one opinion- then we might need to make it configurable

**Comment by lluiscab** [comment]
> I think we need to check if there is "one true way" to do codeowners logic, and if this is it. If the way to do it is opinionated - and the vscode extension is one opinion- then we might need to make it configurable

The vscode extension used the npm library [codeowners](https://www.npmjs.com/package/codeowners) which has similar logic to what was already present on this codebase.

The extra sorting by best match logic has been added to guarantee that the "best" owner fill be chosen when assigning the MR in platforms such as a self hosted GitLab installation which only allow for one assignee and one reviewer per MR

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.spec.ts:44]:
```suggestion
      const mockCodeOwners = codeBlock`
```
use template helper, which strips leading whitespace to ensure proper alignment

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.spec.ts:40]: use codeBlock format instead

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:46]: no, use newline regex

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:45]:
```suggestion
```
shouldn't necessary

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts]: Solved in 8d6197b63848914cd42841eade789282bfca0efd

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts]: Solved in 8d6197b63848914cd42841eade789282bfca0efd

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/code-owners.spec.ts:1]: The maintainers should check if this PR (partially) fixes this issue:

- #12611

Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts:1]: I believe that this PR will fix the issue, as rules are now matched against each file and not against the entire list of modified files as it was done before

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/code-owners.spec.ts:54]:
```suggestion
        # Assign each package to its respective user
```

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/code-owners.spec.ts:51]:
```suggestion
        # Assigning an owner to lockfiles causes problems, because pull requests are assigned to wrong users
```

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/code-owners.spec.ts:164]:
```suggestion
    it('does not require all files to match a single rule, regression test for #12611', async () => {
```

This looks like a regression test, might be good to mention this in the test description?

File: lib/workers/repository/update/pr/code-owners.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/code-owners.ts:70]:
```suggestion
      // Map through all prFiles and match said file(s) with all the rules
```

Maybe the comment needs to mention `file(s)`?

File: lib/workers/repository/update/pr/code-owners.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/code-owners.ts:98]:
```suggestion
        // Match file again but this time only with emptyRules, to ensure that files which have no owner set remain owner-less
```

I think this is easier to understand, but check if it's still technically correct. 😉

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/code-owners.spec.ts:60]: This may just be a "me problem": it's hard to keep track of _who_ owns which file(s).

Can we use more descriptive names for the `packages/*` or the `@users`? This could help make things easier to remember maybe?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some grammar/style fixes for the comments. And a question about the usernames/package names for the test.

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts:60]: um yeah, I just wanted to keep the usernames from the tests that were already there, but we could change the package names and or usernames to have more descriptive names. Any suggestions?

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts:51]: Honestly, that shouldn't be an issue anymore, that comment was copied from a CODEOWNERS file used at work, we had to specifically remove any owners from the `yarn.lock` file as renovate would incorrectly assign pull requests. 

With the new implementation, I think it would actually be good to assign owners to lock-files as a "fallback" for cases where a better match has not been found.

Comment (HonkingGoose) [reply lib/workers/repository/update/pr/code-owners.spec.ts:51]: > Honestly, that shouldn't be an issue anymore, that comment was copied from a `CODEOWNERS `file used at work, we had to specifically remove any owners from the yarn.lock file as renovate would incorrectly assign pull requests.

Whoops, I thought it was a _developer_ comment. It's just a copy/paste of an example `CODEOWNERS `file from a production repository. 🙈 

I'll let you decide what to do with the comment. Though we probably should have a test to confirm we're properly ignoring comments in `CODEOWNERS` files.

> With the new implementation, I think it would actually be good to assign owners to lock-files as a "fallback" for cases where a better match has not been found.

I'll let the maintainers decide. 😉

Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts:51]: > Though we probably should have a test to confirm we're properly ignoring comments in CODEOWNERS 

Test was already there, you don't see it in the PR diff because I didn't need to modify it.

https://github.com/renovatebot/renovate/blob/11adb16758108c19b7c8887e73d848cb990ecaa7/lib/workers/repository/update/pr/code-owners.spec.ts#L205

Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts:51]: > > With the new implementation, I think it would actually be good to assign owners to lock-files as a "fallback" for cases where a better match has not been found.

> I'll let the maintainers decide. wink

I'd say that the maintainers here don't have anything to decide, the final configuration of the CODEOWNERS file is up to the maintainers of the target repository, not of the renovate project. If someone wants to add the line `yarn.lock @someone`, then go ahead, in the majority of cases there's going to be a better match (with a higher score) and this line would act as a fallback (for lock file maintenance, for example)

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.spec.ts:28]: please use codeBlock format instead

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (viceice) [reply lib/workers/repository/update/pr/code-owners.spec.ts:51]: i would prefer to not have the extended description here

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.spec.ts:166]: use codeBlock format

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.spec.ts:184]: same

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:8]: use interface

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:13]: same

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:38]:
```suggestion
    if (!prFiles?.length) {
```
so we stop for empty list too

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:56]: how expensive is `ignore().add(pattern)`? we could create a variable before the return, so it will kept as reference and we skip the initializing on every match call.

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:79]: you should store the score on the rule, so we can directly use and don't need to check the length of the pattern multiple times.

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:110]: i think we should log this as metadata instead of JSON inside the message.

@rarkins WDYT?

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:154]: same here

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts]: Solved in c2888e3faa6c3ff7a3c74f5ea227b8ae4e19a92b

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts]: Solved in c2888e3faa6c3ff7a3c74f5ea227b8ae4e19a92b

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.spec.ts]: Solved in c2888e3faa6c3ff7a3c74f5ea227b8ae4e19a92b

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts]: Solved in 1405529e47463d85a160e801d4993bd3372c1f80

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts]: Solved in b8fa2da6f9879d2a4d7987ba810280634d1068aa

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (HonkingGoose) [reply lib/workers/repository/update/pr/code-owners.spec.ts:54]:
```suggestion
```

@viceice I'm not sure what you mean here:

1. Keep the "old" comment, or
2. Drop the comment completely

If you want option 2 you can apply my suggestion. 😉

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (HonkingGoose) [reply lib/workers/repository/update/pr/code-owners.spec.ts:63]: I'm sorry, this is too complicated for me to figure out. I hope one of the maintainers can help!

File: lib/workers/repository/update/pr/code-owners.ts
Comment (rarkins) [reply lib/workers/repository/update/pr/code-owners.ts:111]: For debug messages I generally prefer no metadata

**Comment by lluiscab** [comment]
Any updates on the status of this PR?

**Comment by rarkins** [comment]
Please get coverage and tests passing and we should be good to go

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:40]: needs coverage

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:74]: please extract the map callbacks to seperate module functions for less code nesting and better readabillity.

**Comment by lluiscab** [comment]
just pushed a new test to cover the only if statement that was not covered (no files in pr), coverage should now pass.

on the other side, the test failures reported don't seem to have any relationship with this PR

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:116]: same here, extract the callback function

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts]: Solved in dbf7502a9541316f3e698d95756a8885f60bd144

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts:129]: Solved in dbf7502a9541316f3e698d95756a8885f60bd144

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:15]: convert to full function declaration

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:41]: same here

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts:15]: you should probably disable arrow functions with some eslint rule

File: lib/workers/repository/update/pr/code-owners.spec.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.spec.ts:39]:
```suggestion
      fs.readLocalFile.mockResolvedValueOnce('* @jimmy');
```

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:45]: please use for-of loop

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:74]: wouldn't it be better to use a map outside of the loops instead of searching in an array so often? (perf)

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts:45]: instead of using `reduce`? is there any reason for that change? imo it's cleaner with reduce

Comment (viceice) [reply lib/workers/repository/update/pr/code-owners.ts:45]: code is cleaner with for-of loop, also reduces nesting.

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [reply lib/workers/repository/update/pr/code-owners.ts:74]: wrong, renovate runs on huge repos which have hundreds of change files. so performance always matters.

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts]: Solved in d2b3880f735ac5fd869d1013aab5207209eeea28

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:137]: did you test how maps are serialized?

Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts]: Just pushed e222e7e76c8f4940f62a979bd658bf00509f16ec with a fix, turns out Maps and stringify don't play along.

**Comment by lluiscab** [comment]
> > Any upates? I'd like to get this merged and released on `whitesource/renovate` soon
> 
> please use the re-request review function

You'll have to tell me how to do that, I see no such button on the UI

**Comment by rarkins** [comment]
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review#

**Review by rarkins — APPROVED** [review]

**Review by rarkins — DISMISSED** [review]

**Review by HonkingGoose — APPROVED** [review]
I'm done with my very surface-level review, I'll let the maintainers review this code properly. 😉

File: lib/workers/repository/update/pr/code-owners.ts
Comment (viceice) [inline lib/workers/repository/update/pr/code-owners.ts:139]:
```suggestion
      `CODEOWNERS matched the following files: ${fileOwners.map((f) => f.file).join(', ')}`
```
isn't this simpler?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/workers/repository/update/pr/code-owners.ts
Comment (lluiscab) [reply lib/workers/repository/update/pr/code-owners.ts:139]: definetly, just kept the JSON stringify as it was there originally (see https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/pr/code-owners.ts#LL50C7-L52C10), we can go ahead and replace it with a string join for simplicity

**Review by rarkins — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Comment by lluiscab** [comment]
Quick question, once this changes are merged, when will they be available in Mend Renovate ([renovate-on-prem](https://github.com/mend/renovate-on-prem/))?

**Comment by rarkins** [comment]
@lluiscab updates in the hosted app are approved manually, typically in <7 days from once the official image is published to Docker Hub

**Pull request merged**
